Support QU and SP break-class in fast path

Constellation opened the pull request:
#### 52e12b92108758547e05df45cc8e8a439b2135df
<pre>
Support QU and SP break-class in fast path
<a href="https://bugs.webkit.org/show_bug.cgi?id=276765">https://bugs.webkit.org/show_bug.cgi?id=276765</a>
<a href="https://rdar.apple.com/131982061">rdar://131982061</a>

Reviewed by Alan Baradlay.

This patch adds QU and SP break-class support in BreakLines.
We intentionally do not categorize u201c u201d as QU since they sometimes are categorized into open/close punctuations.

* Source/WebCore/rendering/BreakLines.h:
(WebCore::BreakLines::nextBreakablePosition):
(WebCore::BreakLines::classify):

Canonical link: <a href="https://commits.webkit.org/281122@main">https://commits.webkit.org/281122@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/546c2ea77c92ee4ef008d6cee59c76e6d99e5785

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/58814 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/38142 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/11304 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/62446 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/9258 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/60943 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/45779 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9458 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/47572 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6590 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/60845 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35688 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/50858 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/28428 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32421 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/8149 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/8262 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54376 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8429 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/64147 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/2727 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8412 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54896 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/2736 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/50885 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54989 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2286 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8777 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/33972 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/35056 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/36141 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/34802 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

fantasai commented on Source/WebCore/rendering/BreakLines.h:
```diff
@@ -178,48 +182,57 @@ inline size_t BreakLines::nextBreakablePosition(CachedLineBreakIteratorFactory&
+                    if (lineBreakTable.unsafeLookup(beforeCharacter, after))
                         return i;
                 } // Else at least one is an ASCII control character; don't break.
                 continue;
```
- When we have before=space, we don't need to do an ascii table lookup, we can just `continue;` and avoid the conditionals after it. (See that it hits the continue clause just below.)
- It's not clear to me that adding a conditional this high up for a relatively rare character is wise, given how sensitive this code seems to be to adding conditionals--this adds a new one for every single character in the document. Instead I would consider capturing this case where we detect kGL in the `// Non-ASCII rapid lookup.` section below.

Constellation replied:
Sounds good. Will add kSP and handle things.

fantasai commented on Source/WebCore/rendering/BreakLines.h:
```diff
@@ -421,6 +439,39 @@ inline BreakLines::BreakClass BreakLines::classify(UChar character)
+    case 0x1980 / 0x80:
+        return kWeird;
+    case 0x2000 / 0x80:
+        if (character == 0x2018 || character == 0x2019)
+            return kQU;
```
I saw you explained that double quotation marks are sometimes categorized as open/close punctuation. Are we sure that single quotations are not also categorized that way? They are usually treated as alternatives to double quotes when the nesting level has increased, so I would expect them to behave the same.

Constellation replied:
> I saw you explained that double quotation marks are sometimes categorized as open/close punctuation. Are we sure that single quotations are not also categorized that way?

Yes, confirmed via ICU rules. They behave differently in some specific locale cases.

fantasai replied:
Yes, but you're sure we do not have that problem with single quotation marks in those locales?

Constellation replied:
Yes, I mean, we ensured that only double quotes are excluded.

fantasai reviewed (changes requested):
What patterns are you trying to capture here? It's hard to evaluate if you're succeeding.

Constellation commented on Source/WebCore/rendering/BreakLines.h:
```diff
@@ -178,48 +182,57 @@ inline size_t BreakLines::nextBreakablePosition(CachedLineBreakIteratorFactory&
             }
             // Handle special cases.
-            if (pair & kGL && !(pair & kWeird)) // Keep nbsp high in our list.
+            if (pair & (kGL | kQU) && !(pair & kWeird)) // Keep nbsp high in our list.
```
Doesn't it handle both cases? It is `kGL | kQU` so this extends the case to `kQU`, but not breaking `kGL` case.

fantasai replied:
Yes, you're right. I got mixed up. :)

fantasai commented on Source/WebCore/rendering/BreakLines.h:
```diff
@@ -190,36 +194,39 @@ inline size_t BreakLines::nextBreakablePosition(CachedLineBreakIteratorFactory&
             int pair = before.type | after.type;
-            if (pair == (kAL | kAL)) {
-                if (words == WordBreakBehavior::BreakAll)
+            // SP+AL SP+QU AL+AL AL+QU QU+QU QU+AL (after's SP is already filtered out).
```
Nit: List the AL + AL combination first, since it's the most important one here. :)

Constellation replied:
Done.

fantasai replied:
I meant in the comment! ^_^;;

fantasai replied:
Ordering of conditionals in this particular function is not nitpicking, it's perf hacking. ;)

fantasai commented on Source/WebCore/rendering/BreakLines.h:
```diff
@@ -190,36 +194,44 @@ inline size_t BreakLines::nextBreakablePosition(CachedLineBreakIteratorFactory&
+            // SP+AL SP+QU AL+AL AL+QU QU+QU QU+AL (after's SP is already filtered out).
+            if ((pair | kSP | kQU) & (kSP | kAL | kQU)) {
+                if constexpr (words == WordBreakBehavior::BreakAll)
+                    return i;
```
I think this needs to be

```suggestion
                if constexpr (words == WordBreakBehavior::BreakAll)
                    if (pair == kAL)
                        return i;
```

Constellation replied:
OK, changed!

fantasai reviewed (approved):
I think we might consider splitting out the kAL and kID conditionals into independent sets for WordBreakBehavior::BreakAll so that for regular text it's

- kAL combo
- kID combo

and for BreakAll it's
- kAL break
- kID combo
- kAL combo

But that's kinda theoretical and not very critical so r+

Constellation committed:
Support QU and SP break-class in fast path
https://bugs.webkit.org/show_bug.cgi?id=276765
rdar://131982061

Reviewed by Alan Baradlay.

This patch adds QU and SP break-class support in BreakLines.
We intentionally do not categorize u201c u201d as QU since they sometimes are categorized into open/close punctuations.

* Source/WebCore/rendering/BreakLines.h:
(WebCore::BreakLines::nextBreakablePosition):
(WebCore::BreakLines::classify):

Canonical link: https://commits.webkit.org/281122@main
